crates/core/src/catalog.rs

```diff
@@ -1,43 +1,61 @@
 use crate::config::Flavor;
-use crate::network::request_async;
+use crate::error;
 use crate::Result;
+use async_std::task;
 use chrono::prelude::*;
+use futures::future::join_all;
+use std::time::Duration;
 
 use isahc::{config::RedirectPolicy, prelude::*};
 use serde::Deserialize;
 
+const DEFAULT_TIMEOUT: u64 = 30;
+
 const CURSE_CATALOG_URL: &str =
     "https://github.com/casperstorm/ajour-catalog/releases/latest/download/curse.json";
 const TUKUI_CATALOG_URL: &str =
     "https://github.com/casperstorm/ajour-catalog/releases/latest/download/tukui.json";
 const WOWI_CATALOG_URL: &str =
     "https://github.com/casperstorm/ajour-catalog/releases/latest/download/wowi.json";
 
-pub async fn get_catalog() -> Result<Catalog> {
+const CATALOG_URLS: [&str; 3] = [WOWI_CATALOG_URL, CURSE_CATALOG_URL, TUKUI_CATALOG_URL];
+
+pub async fn get_catalog_addons_from(url: &str) -> Result<Vec<CatalogAddon>> {
     let client = HttpClient::builder()
         .redirect_policy(RedirectPolicy::Follow)
         .max_connections_per_host(6)
+        .timeout(Duration::from_secs(DEFAULT_TIMEOUT))
         .build()
         .unwrap();
 
-    let mut curse_resp = request_async(&client, CURSE_CATALOG_URL, vec![], Some(30)).await?;
-    let mut tukui_resp = request_async(&client, TUKUI_CATALOG_URL, vec![], Some(30)).await?;
-    let mut wowi_resp = request_async(&client, WOWI_CATALOG_URL, vec![], Some(30)).await?;
-
-    let mut addons = vec![];
-    if curse_resp.status().is_success() {
-        let mut catalog: Catalog = curse_resp.json()?;
-        addons.append(&mut catalog.addons);
+    let request = client.get_async(url.to_string());
+    if let Ok(mut response) = request.await {
+        if let Ok(json) = task::spawn_blocking(move || response.json()).await {
+            log::debug!("Successfully fetched and parsed {}", url);
+            Ok(json)
+        } else {
+            log::debug!("Could not parse {}", url);
+            Err(error!("Could not parse catalog"))
+        }
+    } else {
+        log::debug!("Could not fetch {}", url);
+        Err(error!("Could not fetch catalog"))
     }
+}
 
-    if tukui_resp.status().is_success() {
-        let mut catalog: Catalog = tukui_resp.json()?;
-        addons.append(&mut catalog.addons);
+pub async fn get_catalog() -> Result<Catalog> {
+    let mut futures = vec![];
+    for url in CATALOG_URLS.iter() {
+        futures.push(get_catalog_addons_from(url));
     }
 
-    if wowi_resp.status().is_success() {
-        let mut catalog: Catalog = wowi_resp.json()?;
-        addons.append(&mut catalog.addons);
+    let mut addons = vec![];
+    let results = join_all(futures).await;
+    for api_result in results {
+        match api_result {
+            Ok(c) => addons.append(&mut c.clone()),
+            Err(e) => log::debug!("{}", e),
```

Comment:
Instead of logging the error here (it was already logged under `get_catalog_addons_from`), we should actually return it. We display errors in the GUI so the user has feedback to know something went wrong. Currently this function doesn't actually return any error.

Reply:
Good catch.

Reply:
I didn’t want to change function signature.
This means changing calling site, since the function could only return Ok so far.

Moreover, if one source fails, it doesn’t mean that all failed.
I would return an error only if all sources fails.
The debug here is superfluous though, as it doesn’t have the exact source anyway.

There could be a notification bus which the gui could use to have a notification area maybe. But that’s a much bigger change.

Reply:
What do you mean with ` since the function could only return Ok so far.`.
It's a `Result` so it can return a `ClientError` as well :)

Reply:
Yes, but from the code, it is only returning Ok(). It will never return an error.

So the calling site might just ignore errors. I didn’t checked.

Reply:
Made it so it returns an error only if all sources fails.

Reply:
Good call